portability-transfer/src/main/java/org/dataportabilityproject/transfer/PortabilityTransferServiceProviderRegistry.java

```diff
@@ -0,0 +1,89 @@
+  public PortabilityTransferServiceProviderRegistry(List<String> enabledServices,
+      Map<String, TransferServiceProvider> serviceProviderMap) {
+    ImmutableMap.Builder<String, TransferServiceProvider> serviceProviderBuilder = ImmutableMap
+        .builder();
+    ImmutableSet.Builder<String> supportedImportTypes = ImmutableSet.builder();
+    ImmutableSet.Builder<String> supportedExportTypes = ImmutableSet.builder();
+
+    for (String service : enabledServices) {
+      TransferServiceProvider provider = serviceProviderMap.get(service);
+      Preconditions
+          .checkArgument(provider != null, "TransferServiceProvider not found for [%s]", service);
+
+      List<String> importTypes = provider.getImportTypes();
+      List<String> exportTypes = provider.getExportTypes();
+
+      // Check that each registered service has export if it has import.
+      // We do not allow for import only types
+      for (String type : importTypes) {
+        Preconditions.checkArgument(exportTypes.contains(type),
+            "TransferDataType [%s] is available for import but not export in [%s] TransferServiceProvider",
+            type, service);
+        supportedImportTypes.add(type);
+      }
+
+      supportedExportTypes.addAll(exportTypes);
+      serviceProviderBuilder.put(service, provider);
+    }
+
+    this.serviceProviderMap = serviceProviderBuilder.build();
+    this.supportedExportTypes = supportedExportTypes.build();
+    this.supportedImportTypes = supportedImportTypes.build();
+  }
+
+  /**
+   * Returns the exporter that supports the serviceId and transferDataType.
+   *
+   * @param serviceId the service id
+   * @param transferDataType the transfer data type
+   */
+  @Override
+  public Exporter<AuthData, DataModel> getExporter(String serviceId,
+      String transferDataType) {
+    Preconditions.checkArgument(supportedExportTypes.contains(transferDataType),
+        "TransferDataType [%s] is not valid for export", transferDataType);
+    TransferServiceProvider serviceProvider = serviceProviderMap.get(serviceId);
+    Preconditions.checkArgument(serviceProvider != null);
+    return (Exporter<AuthData, DataModel>) serviceProvider.getExporter(transferDataType);
+  }
+
+  /**
+   * Returns the exporter that supports the serviceId and transferDataType.
+   *
+   * @param serviceId the service id
+   * @param transferDataType the transfer data type
+   */
+  @Override
+  public Importer<AuthData, DataModel> getImporter(String serviceId,
+      String transferDataType) {
+    Preconditions.checkArgument(supportedImportTypes.contains(transferDataType),
+        "TransferDataType [%s] is not valid for import", transferDataType);
+    TransferServiceProvider serviceProvider = serviceProviderMap.get(serviceId);
+    Preconditions.checkArgument(serviceProvider != null);
+    return (Importer<AuthData, DataModel>) serviceProvider.getImporter(transferDataType);
```

Comment:
I wasnt sure what the best way to cast the Importer<?, ?> object thats returned from the different service providers. Is this something thats commonly done?

Reply:
We lose the subtype info by using the base classes in the cast and method signature, e.g. AuthData and DataModel.  Can we just return wildcards as well?

Reply:
Done - I had to make some changes to the PortabilityCopier to reflect the new changes and handle the wildcard types. I think I did them correctly after a couple of different casting tries.